pkg/dynamicaccess/controller_test.go

```diff
@@ -1,98 +1,109 @@
+	al.AddPublisher(ctx, kvs1, publisher)
+	kvs1Ref, _ := kvs1.Save(ctx)
+	kvs2 := kvs.New(ls, swarm.ZeroAddress)
+	al.AddGrantee(ctx, kvs2, publisher, &pk2.PublicKey, nil)
+	al.AddPublisher(ctx, kvs2, publisher)
+	kvs2Ref, _ := kvs2.Save(ctx)
+	firstTime := time.Date(1994, time.April, 1, 0, 0, 0, 0, time.UTC).Unix()
+	secondTime := time.Date(2000, time.April, 1, 0, 0, 0, 0, time.UTC).Unix()
+	thirdTime := time.Date(2015, time.April, 1, 0, 0, 0, 0, time.UTC).Unix()
+
+	h.Add(ctx, kvs0Ref, &thirdTime)
+	h.Add(ctx, kvs1Ref, &firstTime)
+	h.Add(ctx, kvs2Ref, &secondTime)
+	return h.Store(ctx)
 }
 
-func TestDecrypt(t *testing.T) {
-	pk := getPrivateKey()
-	ak := encryption.Key([]byte("cica"))
-
-	si := dynamicaccess.NewDefaultSession(pk)
-	aek, _ := si.Key(&pk.PublicKey, [][]byte{{0}, {1}})
-	e2 := encryption.New(aek[1], 0, uint32(0), hashFunc)
-	peak, _ := e2.Encrypt(ak)
-
-	h := mockTestHistory(aek[0], peak)
-	al := setupAccessLogic(pk)
-	gm := dynamicaccess.NewGranteeManager(al)
-	c := dynamicaccess.NewController(h, gm, al)
-	eref, ref := prepareEncryptedChunkReference(ak)
-	// ech := al.EncryptRef(ch, "tag")
+// TODO: separate up down test with fixture, now these just check if the flow works at all
+func TestController_NewUploadDownload(t *testing.T) {
+	ctx := context.Background()
+	publisher := getPrivKey(1)
+	diffieHellman := dynamicaccess.NewDefaultSession(publisher)
+	al := dynamicaccess.NewLogic(diffieHellman)
+	c := dynamicaccess.NewController(ctx, al, mockStorer.ChunkStore(), mockStorer.Cache())
+	ref := swarm.RandAddress(t)
+	_, hRef, encryptedRef, err := c.UploadHandler(ctx, ref, &publisher.PublicKey, nil)
+	assert.NoError(t, err)
+	dref, err := c.DownloadHandler(ctx, time.Now().Unix(), encryptedRef, &publisher.PublicKey, hRef)
+	assert.NoError(t, err)
+	assert.Equal(t, ref, dref)
+}
 
-	ts := int64(0)
-	addr, err := c.DownloadHandler(ts, eref, &pk.PublicKey, "tag")
-	if err != nil {
-		t.Fatalf("DownloadHandler() returned an error: %v", err)
-	}
-	if !addr.Equal(ref) {
-		t.Fatalf("Decrypted chunk address: %s is not the expected: %s", addr, ref)
-	}
+func TestController_ExistingUploadDownload(t *testing.T) {
+	ls := createLs()
+	ctx := context.Background()
+	publisher := getPrivKey(0)
+	diffieHellman := dynamicaccess.NewDefaultSession(publisher)
+	al := dynamicaccess.NewLogic(diffieHellman)
+	c := dynamicaccess.NewController(ctx, al, mockStorer.ChunkStore(), mockStorer.Cache())
+	ref := swarm.RandAddress(t)
+	hRef, err := getHistoryFixture(ctx, ls, al, &publisher.PublicKey)
+	assert.NoError(t, err)
+	_, hRef, encryptedRef, err := c.UploadHandler(ctx, ref, &publisher.PublicKey, &hRef)
+	assert.NoError(t, err)
+	dref, err := c.DownloadHandler(ctx, time.Now().Unix(), encryptedRef, &publisher.PublicKey, hRef)
+	assert.NoError(t, err)
+	assert.Equal(t, ref, dref)
 }
 
-func TestEncrypt(t *testing.T) {
-	pk := getPrivateKey()
-	ak := encryption.Key([]byte("cica"))
+func TestControllerGrant(t *testing.T) {
+	// al := setupAccessLogic()
```

Comment:
The commented fields will still be needed for the final tests? If not, they could be removed.

Reply:
removed